File: 0139_WordBreak/WordBreak.py
from collections import Counter

class Solution1:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        dp=[False]*(len(s)+1)
        dp[0]=True
        Dict = Counter(wordDict)
        for i in range(len(s)):
            for j in range(0,i+1):
                if s[j:i+1] in Dict and dp[j]:
                    dp[i+1]=True
                    break

        return dp[-1]

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        if not wordDict:
            return False
        n = len(s)
        
        result = [False] * (n + 1)
        result[0] = True
        
        for i in range(n):
            if not result[i]:
                continue
            for w in wordDict:
                end = i + len(w)
                if end <= n and not result[end] and s[i: end] == w:
                    result[end] = True
        return result[-1]

File: 0139_WordBreak/WordBreak.cxx
#include<vector>
#include<string>
#include<unordered_set>
#include<unordered_map>

using namespace std;

class Solution {
public:
    
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(), wordDict.end());
        unordered_set<char> checkdict;
        for(auto & word:wordDict){
            for(auto & ch:word){
                checkdict.insert(ch);
            }
        }
        for(auto & ch:s){
            if(!checkdict.count(ch)) return false;
        }
        unordered_map<string, bool> memo;
        return wordBreak(s, dict, memo);
    }
    bool wordBreak( string s, 
                    unordered_set<string> &dict, 
                    unordered_map<string, bool> &memo){
        auto found = memo.find(s);
        if( found != memo.end()){
            return found->second;
        }
        for( int i=1; i<=s.size(); ++i){
            if(dict.count(s.substr(0,i))){
                if(i == s.size()) return true;
                string subStr = s.substr(i);
                if(wordBreak(subStr, dict, memo)){
                    return true;
                } 
                else{
                    memo[subStr]=false;
                }
            }
        }
        memo[s]=false;
        return false;
    }
    
};